internal/worker/worker.go

```diff
@@ -0,0 +1,300 @@
+}
+
+type worker struct {
+	ip     ImagePuller
+	logger logr.Logger
+	mr     ModprobeRunner
+}
+
+func NewWorker(ip ImagePuller, mr ModprobeRunner, logger logr.Logger) Worker {
+	return &worker{
+		ip:     ip,
+		logger: logger,
+		mr:     mr,
+	}
+}
+
+func (w *worker) LoadKmod(ctx context.Context, cfg *kmmv1beta1.ModuleConfig) error {
+	imageName := cfg.ContainerImage
+
+	w.logger.Info("Pulling image", "name", imageName)
+
+	pr, err := w.ip.PullToDisk(ctx, imageName, cfg.InsecurePull)
+	if err != nil {
+		return fmt.Errorf("could not pull %q: %v", imageName, err)
+	}
+
+	if inTree := cfg.InTreeModuleToRemove; inTree != "" {
+		w.logger.Info("Unloading in-tree module", "name", inTree)
+
+		if err = w.mr.Run(ctx, "-rv", inTree); err != nil {
+			return fmt.Errorf("could not remove in-tree module %s: %v", inTree, err)
+		}
+	}
+
+	// TODO copy firmware
+	// TODO handle ModulesLoadingOrder
+
+	moduleName := cfg.Modprobe.ModuleName
+
+	var args []string
+
+	if cfg.Modprobe.RawArgs != nil {
+		args = cfg.Modprobe.RawArgs.Load
+	} else {
+		args = []string{"-vd", filepath.Join(pr.fsDir, cfg.Modprobe.DirName)}
+
+		if cfg.Modprobe.Args != nil {
+			args = append(args, cfg.Modprobe.Args.Load...)
+		}
+
+		args = append(args, moduleName)
+		args = append(args, cfg.Modprobe.Parameters...)
+	}
+
+	return w.mr.Run(ctx, args...)
+}
+
+func (w *worker) ReloadKmod(ctx context.Context, oldCfg, newCfg *kmmv1beta1.ModuleConfig) error {
+	return errors.New("not implemented")
+}
+
+func (w *worker) UnloadKmod(ctx context.Context, cfg *kmmv1beta1.ModuleConfig) error {
+	imageName := cfg.ContainerImage
+
+	w.logger.Info("Pulling image", "name", imageName)
+
+	pr, err := w.ip.PullToDisk(ctx, imageName, cfg.InsecurePull)
+	if err != nil {
+		return fmt.Errorf("could not pull %q: %v", imageName, err)
+	}
+
+	moduleName := cfg.Modprobe.ModuleName
+
+	var args []string
+
+	if cfg.Modprobe.RawArgs != nil {
+		args = cfg.Modprobe.RawArgs.Unload
+	} else {
+		args = []string{"-rvd", filepath.Join(pr.fsDir, cfg.Modprobe.DirName)}
+
+		if cfg.Modprobe.Args != nil {
+			args = append(args, cfg.Modprobe.Args.Unload...)
+		}
+
+		args = append(args, moduleName)
+	}
+
+	w.logger.Info("Unloading module", "name", moduleName)
+
+	if err = w.mr.Run(ctx, args...); err != nil {
+		return fmt.Errorf("could not unload module %s: %v", moduleName, err)
+	}
+
+	// TODO remove firmware
+
+	return nil
+}
+
+type pullResult struct {
+	fsDir string
+}
+
+//go:generate mockgen -source=worker.go -package=worker -destination=mock_worker.go
+
+type ImagePuller interface {
+	PullToDisk(ctx context.Context, imageName string, insecurePull bool) (pullResult, error)
+}
+
+type imagePuller struct {
+	baseDir string
+	logger  logr.Logger
+}
+
+func NewImagePuller(baseDir string, logger logr.Logger) ImagePuller {
+	return &imagePuller{
+		baseDir: baseDir,
+		logger:  logger,
+	}
+}
+
+func (i *imagePuller) PullToDisk(ctx context.Context, imageName string, insecurePull bool) (pullResult, error) {
```

Comment:
General remark: how about we stash the digest implementation for now? We can implement a simple pull onto the container FS. Once we have a working scenario we can come back to it and see if it worth to mount host path and implement the cache with digest. WDYT?

Reply:
The digest implementation saves an image pull and extraction when we need to reload a kmod from the same kmod image. I agree that it's not so useful now, but since it's already there, would you mind keeping it?

Reply:
it is only useful in case we start mounting the host directory into the Pod from the beginning?  Are we going to do it from the beginning? It als means implementing garbage collection on the host. I think this is something that we need to implement (if at all) only at later stage, as an optimization.  If you want to keep the code, then i have no objection, lets just add a comment do describe the digest mechanism and add a TODO to remove it later if needed

Reply:
It is useful without mounting the host directory in the pod, if you use the `kmod reload` feature where you might be pulling the same image twice.

Reply:
but in that case isn't it easier just to check the imageName in the reloadKmod command? It is safer and quicker than trying to write 2 files instead of one (digest). Also saves a failing the pod when the only thing that failed is figest write, while the image extraction was succesfull

Reply:
Now removing `dstDir` (containing `fs` and `digest`) if `digest` does not exist or does not contain the expected value.
Keeping `digest` allows simple retries by calling `LoadKmod` or `UnloadKmod` in a loop. Because `digest` is written only if the extraction was successful, we now on the next try that we have an extracted filesystem we can safely use and skip the pull.

Reply:
since reload command is not relevant anymore, do we still want this optimization in the container FS?

Reply:
I think we do, because it makes retries much easier.

Reply:
only in case we mount a host directory, but that is not what we are doing right now, no?